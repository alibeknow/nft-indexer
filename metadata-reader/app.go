package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
