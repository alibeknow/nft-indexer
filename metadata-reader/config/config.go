package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
