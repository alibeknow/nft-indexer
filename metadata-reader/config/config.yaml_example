server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
