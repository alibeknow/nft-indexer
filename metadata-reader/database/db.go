package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
