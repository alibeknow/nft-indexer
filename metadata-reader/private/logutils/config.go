package logutils

type Config struct {
	Level string
}
