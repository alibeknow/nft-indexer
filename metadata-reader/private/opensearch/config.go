package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
