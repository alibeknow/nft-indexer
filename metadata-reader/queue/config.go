package queue

import "fmt"

type Config struct {
	Protocol string
	Username string
	Password string
	Host     string
	Port     int
	Name     string
	Workers  int
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	return url
}
