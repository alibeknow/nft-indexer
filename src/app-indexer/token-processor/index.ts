export * from './token-processor.controller';
export * from './token-processor.dtos';
export * from './token-processor.module';
export * from './token-processor.provider';
export * from './token-processor.service';
