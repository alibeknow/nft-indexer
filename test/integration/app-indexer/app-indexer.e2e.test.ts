import { v4 as uuid } from 'uuid';

process.env.RABBIT_BLOCK_READ_QUEUE = `test-queue-${uuid()}`;

import { ethers } from 'ethers';
import { setTimeout } from 'timers/promises';
import { Test } from '@nestjs/testing';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { INestApplication } from '@nestjs/common';
import { ClientProxy } from '@nestjs/microservices';
import { IIndexerConfig, IServiceConfig, getIndexerConfig } from '@app-indexer/app.config';
import { TokenProcessorModule } from '@app-indexer/token-processor';
import * as nfts from '@shared/nfts';
import { DB, DBClass, IDatabaseConfig, MongoSort, getDatabaseConfig } from '@shared/db';
import { ServiceEvents, bootstrapMicroservice, clientFactory } from '@shared/microservices';
import { fixtureEvents } from './fixtures';

jest.mock('@shared/nfts', () => {
  const original = jest.requireActual('@shared/nfts');

  return {
    ...original,
    getContractInstance: jest.fn(),
  };
});

jest.spyOn(nfts, 'getContractInstance').mockImplementation((...args) => {
  return {
    address: args[0] as string,
    name: () => new Promise((resolve) => resolve(args[0])),
    tokenURI: (id: string) => new Promise((resolve) => resolve(`https://tokenUri/${id}`)),
    uri: (id: string) => new Promise((resolve) => resolve(`https://uri/${id}`)),
  } as unknown as ethers.Contract;
});

jest.setTimeout(100000);

describe('[app-indexer] (e2e)', () => {
  let app: INestApplication;
  let client: ClientProxy;
  let db: DBClass;

  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `app-indexer-e2e-test-${dbConfig.db}` });
    await db.open();

    await Promise.all([
      db.events().deleteMany({}),
      db.tokens().deleteMany({}),
      db.contracts().deleteMany({}),
    ]);
    await db.events().insertMany(fixtureEvents);

    const moduleRef = await Test.createTestingModule({
      imports: [
        TokenProcessorModule,
        ConfigModule.forRoot({
          load: [ getIndexerConfig ],
          isGlobal: true,
        }),
      ],
    })
      .overrideProvider(DB)
      .useValue(db)
      .compile();

    app = moduleRef.createNestApplication();

    const configService = moduleRef.get<ConfigService<IIndexerConfig>>(ConfigService<IIndexerConfig>);
    const serviceConfig = configService.get<IServiceConfig>('service') as IServiceConfig;

    await bootstrapMicroservice(app, 'blocksQueue');
    await app.listen(serviceConfig.port);

    client = clientFactory(configService, 'blocksQueue');
    await client.connect();
  });

  afterEach(async () => {
    await Promise.all([
      db.tokens().deleteMany({}),
      db.contracts().deleteMany({}),
    ]);
  });

  afterAll(async () => {
    await db.events().deleteMany({});
    await db.close(true);
    await app.close();
    await setTimeout(1000);
    await client.close();
  });

  it('emit event to the app-indexer', async () => {
    client.emit(
      ServiceEvents.INDEX_BLOCK_RANGE,
      { blockFrom: 14316460, blockTo: 14316461, reread: false },
    );

    await setTimeout(5000);

    const contractsResult = await db.contracts()
      .find({})
      .project({ createdAt: 0, updatedAt: 0 })
      .sort({ _id: MongoSort.ASC })
      .toArray();

    const tokensResult = await db.tokens()
      .find({})
      .project({ createdAt: 0, updatedAt: 0 })
      .sort({ _id: MongoSort.ASC })
      .toArray();

    expect(tokensResult).toMatchSnapshot();
    expect(contractsResult).toMatchSnapshot();
  });
});
