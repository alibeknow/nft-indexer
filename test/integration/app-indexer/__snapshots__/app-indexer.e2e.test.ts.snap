// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[app-indexer] (e2e) emit event to the app-indexer 1`] = `
[
  {
    "_id": "eth:0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c:0x09",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0x33EeCbf908478C10614626A9D304bfe18B78DD73:0x74727e18",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85:0x236e3738237711e27d68d047bd63904459456f82125b1b930687ba3498edc469",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017d",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f:0x1ecb",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0xC36442b4a4522E871399CD717aBDD847Ab11FE88:0x030aee",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0xa7d8d9ef8D8Ce8992Df33D8b8CF4Aebabd5bD270:0x10272239",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0xcCF200D7694a7f0a325c8AFD905c997a6Cc3Ee10:0x01",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
  {
    "_id": "eth:0xcCF200D7694a7f0a325c8AFD905c997a6Cc3Ee10:0x04",
    "block": 14316461,
    "count": 1,
    "from": "0x0000000000000000000000000000000000000000",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "",
  },
]
`;

exports[`[app-indexer] (e2e) emit event to the app-indexer 2`] = `
[
  {
    "_id": "eth:0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c",
    "address": "0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-721",
  },
  {
    "_id": "eth:0x33EeCbf908478C10614626A9D304bfe18B78DD73",
    "address": "0x33EeCbf908478C10614626A9D304bfe18B78DD73",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-721",
  },
  {
    "_id": "eth:0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85",
    "address": "0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-721",
  },
  {
    "_id": "eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5",
    "address": "0x74ba9caaa5847e142263d5F7645fFeD92e5087E5",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-721",
  },
  {
    "_id": "eth:0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
    "address": "0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-721",
  },
  {
    "_id": "eth:0xC36442b4a4522E871399CD717aBDD847Ab11FE88",
    "address": "0xC36442b4a4522E871399CD717aBDD847Ab11FE88",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-721",
  },
  {
    "_id": "eth:0xa7d8d9ef8D8Ce8992Df33D8b8CF4Aebabd5bD270",
    "address": "0xa7d8d9ef8D8Ce8992Df33D8b8CF4Aebabd5bD270",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-721",
  },
  {
    "_id": "eth:0xcCF200D7694a7f0a325c8AFD905c997a6Cc3Ee10",
    "address": "0xcCF200D7694a7f0a325c8AFD905c997a6Cc3Ee10",
    "block": 14316461,
    "collectionImage": null,
    "name": "",
    "type": "ERC-1155",
  },
]
`;
